wandb/integration/openai/fine_tune.py

```diff
@@ -0,0 +1,352 @@
+        for _, row in df_results.iterrows():
+            metrics = {k: v for k, v in row.items() if not np.isnan(v)}
+            step = metrics.pop("step")
+            if step is not None:
+                step = int(step)
+            wandb.log(metrics, step=step)
+        fine_tuned_model = fine_tune.fine_tuned_model
+        if fine_tuned_model is not None:
+            wandb.summary["fine_tuned_model"] = fine_tuned_model
+
+        # training/validation files and fine-tune details
+        cls._log_artifacts(fine_tune, project, entity)
+
+        # mark run as complete
+        wandb.summary["status"] = "succeeded"
+
+        wandb.finish()
+        return True
+
+    @classmethod
+    def _ensure_logged_in(cls):
+        if not cls._logged_in:
+            if wandb.login():
+                cls._logged_in = True
+            else:
+                raise Exception("You need to log in to wandb")
+
+    @classmethod
+    def _get_wandb_run(cls, run_path):
+        cls._ensure_logged_in()
+        try:
+            if cls._wandb_api is None:
+                cls._wandb_api = wandb.Api()
+            return cls._wandb_api.run(run_path)
+        except Exception:
+            return None
+
+    @classmethod
+    def _get_wandb_artifact(cls, artifact_path):
+        cls._ensure_logged_in()
+        try:
+            if cls._wandb_api is None:
+                cls._wandb_api = wandb.Api()
+            return cls._wandb_api.artifact(artifact_path)
+        except Exception:
+            return None
+
+    @classmethod
+    def _get_config(cls, fine_tune):
+        config = dict(fine_tune)
+        config["result_files"] = config["result_files"][0]
+        if config.get("created_at"):
+            config["created_at"] = datetime.datetime.fromtimestamp(config["created_at"])
+        return config
+
+    @classmethod
+    def _log_artifacts(cls, fine_tune, project, entity):
+        # training/validation files
+        training_file = fine_tune.training_file if fine_tune.training_file else None
+        validation_file = (
+            fine_tune.validation_file if fine_tune.validation_file else None
+        )
+        for file, prefix, artifact_type in (
+            (training_file, "train", "training_files"),
+            (validation_file, "valid", "validation_files"),
+        ):
+            if file is not None:
+                cls._log_artifact_inputs(file, prefix, artifact_type, project, entity)
+
+        # fine-tune details
+        fine_tune_id = fine_tune.id
+        artifact = wandb.Artifact(
+            "model_metadata",
+            type="model",
+            metadata=dict(fine_tune),
+        )
+        with artifact.new_file("model_metadata.json", mode="w", encoding="utf-8") as f:
+            dict_fine_tune = dict(fine_tune)
+            dict_fine_tune["hyperparameters"] = dict(dict_fine_tune["hyperparameters"])
+            json.dump(dict_fine_tune, f, indent=2)
+        wandb.run.log_artifact(
```

Comment:
should you check if there is a run?

Reply:
I don't think there is a need to check for a run here. if it is a unique fine-tune job, a new wandb run will be initialised by the logger. but I can put a check anyway.